// @sanity/form-builder/lib/inputs/Array/ItemValue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ItemValue = require('./styles/ItemValue.css');

var _ItemValue2 = _interopRequireDefault(_ItemValue);

var _ConfirmButton = require('./ConfirmButton');

var _ConfirmButton2 = _interopRequireDefault(_ConfirmButton);

var _linkIcon = require('part:@sanity/base/link-icon');

var _linkIcon2 = _interopRequireDefault(_linkIcon);

var _fold = require('part:@sanity/components/edititem/fold');

var _fold2 = _interopRequireDefault(_fold);

var _popover = require('part:@sanity/components/edititem/popover');

var _popover2 = _interopRequireDefault(_popover);

var _fullscreen = require('part:@sanity/components/dialogs/fullscreen');

var _fullscreen2 = _interopRequireDefault(_fullscreen);

var _default = require('part:@sanity/components/dialogs/default');

var _default2 = _interopRequireDefault(_default);

var _FormBuilderInput = require('../../FormBuilderInput');

var _PatchEvent = require('../../PatchEvent');

var _PatchEvent2 = _interopRequireDefault(_PatchEvent);

var _Preview = require('../../Preview');

var _Preview2 = _interopRequireDefault(_Preview);

var _sortable = require('part:@sanity/components/lists/sortable');

var _router = require('part:@sanity/base/router');

var _resolveTypeName = require('../../utils/resolveTypeName');

var _pathUtils = require('../../utils/pathUtils');

var PathUtils = _interopRequireWildcard(_pathUtils);

var _barsIcon = require('part:@sanity/base/bars-icon');

var _barsIcon2 = _interopRequireDefault(_barsIcon);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*:: import type {ArrayType, ItemValue} from './typedefs'*/
/*:: import type {Node} from 'react'*/
/*:: import type {Path} from '../../typedefs/path'*/
/*:: import type {Type} from '../../typedefs'*/


const DragHandle = (0, _sortable.createDragHandle)(() => _react2.default.createElement(
  'span',
  { className: _ItemValue2.default.dragHandle },
  _react2.default.createElement(_barsIcon2.default, null)
));

const DIALOG_ACTIONS = [{
  index: '1',
  name: 'close',
  title: 'Close'
}, {
  index: '2',
  name: 'delete',
  kind: 'simple',
  title: 'Delete',
  color: 'danger',
  secondary: true
}];

/*:: type Props = {
  type: ArrayType,
  value: ItemValue,
  level: number,
  layout?: 'media' | 'default',
  onRemove: (ItemValue) => void,
  onChange: (PatchEvent, ItemValue) => void,
  onFocus: (Path) => void,
  onBlur: void => void,
  focusPath: Path
}*/


function pathSegmentFrom(value) {
  return { _key: value._key };
}

function hasFocusInPath(path, value) {
  return path.length === 1 && PathUtils.isSegmentEqual(path[0], pathSegmentFrom(value));
}

class RenderItemValue extends _react2.default.Component /*:: <Props>*/ {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.handleEditStart = event => {
      this.setFocus([PathUtils.FOCUS_TERMINATOR]);
    }, this.handleFocus = () => {
      this.setFocus();
    }, this.handleEditStop = () => {
      this.setFocus();
    }, this.handleKeyPress = (event /*: SyntheticKeyboardEvent<*>*/) => {
      if (event.key === 'Enter' || event.key === ' ') {
        event.preventDefault();
        this.setFocus([PathUtils.FOCUS_TERMINATOR]);
      }
    }, this.handleRemove = () => {
      var _props = this.props;
      const onRemove = _props.onRemove,
            value = _props.value;

      onRemove(value);
    }, this.handleChange = (event /*: PatchEvent*/) => {
      var _props2 = this.props;
      const onChange = _props2.onChange,
            value = _props2.value;

      onChange(event, value);
    }, this.setFocusArea = (el /*: ?FocusArea*/) => {
      this._focusArea = el;
    }, this.handleDialogAction = action => {
      if (action.name === 'close') {
        this.handleEditStop();
      }
      if (action.name === 'delete') {
        // Needs a proper confirm dialog later
        if (window.confirm('Do you really want to delete?')) {
          // eslint-disable-line no-alert
          this.handleRemove();
        }
      }
    }, _temp;
  }

  componentDidMount() {
    var _props3 = this.props;
    const focusPath = _props3.focusPath,
          value = _props3.value;

    if (hasFocusInPath(focusPath, value)) {
      this.focus();
    }
  }

  componentDidUpdate(prevProps) {
    const hadFocus = hasFocusInPath(prevProps.focusPath, prevProps.value);
    const hasFocus = hasFocusInPath(this.props.focusPath, this.props.value);
    if (!hadFocus && hasFocus) {
      this.focus();
    }
  }

  getMemberType() /*: ?Type*/ {
    var _props4 = this.props;
    const value = _props4.value,
          type = _props4.type;

    const itemTypeName = (0, _resolveTypeName.resolveTypeName)(value);
    return type.of.find(memberType => memberType.name === itemTypeName);
  }

  setFocus(path /*: Path*/ = []) {
    var _props5 = this.props;
    const value = _props5.value,
          onFocus = _props5.onFocus;

    onFocus([{ _key: value._key }, ...path]);
  }

  focus() {
    if (this._focusArea) {
      this._focusArea.focus();
    }
  }

  renderEditItemForm(item /*: ItemValue*/) /*: Node*/ {
    var _props6 = this.props;
    const type = _props6.type,
          focusPath = _props6.focusPath,
          onFocus = _props6.onFocus,
          onBlur = _props6.onBlur;

    const options = type.options || {};

    const memberType = this.getMemberType() || {};

    const content = _react2.default.createElement(_FormBuilderInput.FormBuilderInput, {
      type: memberType,
      level: 0,
      value: item,
      onChange: this.handleChange,
      onFocus: onFocus,
      onBlur: onBlur,
      focusPath: focusPath,
      path: [{ _key: item._key }]
    });

    // test focus issues by uncommenting the next line
    // return content

    if (options.editModal === 'fullscreen') {
      return _react2.default.createElement(
        _fullscreen2.default,
        { title: memberType.title, onClose: this.handleEditStop, isOpen: true },
        content
      );
    }

    if (options.editModal === 'fold') {
      return _react2.default.createElement(
        'div',
        { className: _ItemValue2.default.popupAnchorRelative },
        _react2.default.createElement(
          _fold2.default,
          { title: memberType.title, onClose: this.handleEditStop },
          content
        )
      );
    }

    if (options.editModal === 'popover') {
      return _react2.default.createElement(
        'div',
        { className: _ItemValue2.default.popupAnchor },
        _react2.default.createElement(
          _popover2.default,
          { onClose: this.handleEditStop, key: item._key },
          content
        )
      );
    }

    return _react2.default.createElement(
      'div',
      { className: _ItemValue2.default.popupAnchor },
      _react2.default.createElement(
        _default2.default,
        {
          isOpen: true,
          onClose: this.handleEditStop,
          key: item._key,
          title: 'Edit',
          onAction: this.handleDialogAction,
          showCloseButton: false,
          actions: DIALOG_ACTIONS
        },
        _react2.default.createElement(
          'div',
          { className: _ItemValue2.default.defaultDialogContent },
          content
        )
      )
    );
  }

  renderItem() {
    var _props7 = this.props;
    const value = _props7.value,
          type = _props7.type;

    const options = type.options || {};
    const isGrid = options.layout === 'grid';
    const isSortable = options.sortable !== false;
    const previewLayout = isGrid ? 'media' : 'default';

    return _react2.default.createElement(
      'div',
      { className: _ItemValue2.default.inner },
      !isGrid && isSortable && _react2.default.createElement(DragHandle, null),
      _react2.default.createElement(
        'div',
        {
          tabIndex: 0,
          onClick: this.handleEditStart,
          onKeyPress: this.handleKeyPress,
          className: _ItemValue2.default.previewWrapper
        },
        _react2.default.createElement(
          'div',
          {
            tabIndex: -1,
            ref: this.setFocusArea,
            className: _ItemValue2.default.previewWrapperHelper,
            onFocus: this.handleFocus
          },
          _react2.default.createElement(_Preview2.default, {
            layout: previewLayout,
            value: value,
            type: this.getMemberType()
          })
        )
      ),
      _react2.default.createElement(
        'div',
        { className: _ItemValue2.default.functions },
        value._ref && _react2.default.createElement(
          _router.IntentLink,
          {
            className: _ItemValue2.default.linkToReference,
            intent: 'edit',
            params: { id: value._ref }
          },
          _react2.default.createElement(_linkIcon2.default, null)
        ),
        !type.readOnly && _react2.default.createElement(_ConfirmButton2.default, {
          title: 'Remove this item',
          onConfirm: this.handleRemove
        })
      )
    );
  }

  render() {
    var _props8 = this.props;
    const value = _props8.value,
          focusPath = _props8.focusPath,
          type = _props8.type;


    const options = type.options || {};
    const isGrid = options.layout === 'grid';
    const isExpanded = PathUtils.isExpanded(value, focusPath);

    return _react2.default.createElement(
      'div',
      {
        className: isGrid ? _ItemValue2.default.gridItem : _ItemValue2.default.listItem,
        ref: this.setElement
      },
      this.renderItem(),
      _react2.default.createElement(
        'div',
        {
          className: options.editModal === 'fold' ? _ItemValue2.default.editRootFold : _ItemValue2.default.editRoot
        },
        isExpanded && this.renderEditItemForm(value)
      )
    );
  }
}
exports.default = RenderItemValue;
RenderItemValue.defaultProps = {
  level: 0
};
