// @sanity/components/lib/utilities/ActivateOnFocus.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ActivateOnFocus = require('./styles/ActivateOnFocus.css');

var _ActivateOnFocus2 = _interopRequireDefault(_ActivateOnFocus);

var _reactClickOutside = require('react-click-outside');

var _reactClickOutside2 = _interopRequireDefault(_reactClickOutside);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ActivateOnFocus extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.state = {
      hasFocus: false
    }, this.setEventHandlerElement = element => {
      this._eventHandlerElement = element;
    }, this.handleClick = event => {
      if (!this.state.hasFocus) {
        this.setState({
          hasFocus: true
        });
        this.props.onFocus();
      }
    }, this.handleClickOutside = () => {
      if (this.state.hasFocus) {
        this.setState({
          hasFocus: false
        });
        this.props.onBlur();
      }
    }, this.handleFocus = event => {
      this.setState({
        hasFocus: true
      });
      this.props.onFocus();
    }, this.handleBlur = event => {
      const enableBlur = this.props.enableBlur;

      if (enableBlur) {
        this.setState({
          hasFocus: false
        });
      }
      this.props.onBlur();
    }, _temp;
  }

  render() {
    var _props = this.props;
    const message = _props.message,
          children = _props.children,
          isActive = _props.isActive;
    const hasFocus = this.state.hasFocus;


    if (isActive) {
      return children;
    }

    return _react2.default.createElement(
      'div',
      {
        className: hasFocus ? _ActivateOnFocus2.default.hasFocus : _ActivateOnFocus2.default.noFocus,
        onFocus: this.handleFocus,
        onBlur: this.handleBlur
      },
      _react2.default.createElement(
        'div',
        {
          className: _ActivateOnFocus2.default.eventHandler,
          onClick: this.handleClick
        },
        _react2.default.createElement('div', { className: _ActivateOnFocus2.default.overlay }),
        _react2.default.createElement(
          'div',
          { className: _ActivateOnFocus2.default.message },
          message
        )
      ),
      _react2.default.createElement(
        'div',
        { className: _ActivateOnFocus2.default.content },
        children
      )
    );
  }
}

ActivateOnFocus.propTypes = {
  children: _propTypes2.default.node.isRequired,
  message: _propTypes2.default.string,
  isActive: _propTypes2.default.bool,
  enableBlur: _propTypes2.default.bool,
  onFocus: _propTypes2.default.func,
  onBlur: _propTypes2.default.func
};
ActivateOnFocus.defaultProps = {
  enableBlur: true,
  message: 'Click to activate…',
  isActive: false,
  onFocus() {},
  onBlur() {}
};
exports.default = (0, _reactClickOutside2.default)(ActivateOnFocus);
