// @sanity/form-builder/lib/sanity/SanityFormBuilder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _SanityFormBuilderContext = require('./SanityFormBuilderContext');

var _SanityFormBuilderContext2 = _interopRequireDefault(_SanityFormBuilderContext);

var _FormBuilderInput = require('../FormBuilderInput');

var _SimpleFocusManager = require('./focusManagers/SimpleFocusManager');

var _SimpleFocusManager2 = _interopRequireDefault(_SimpleFocusManager);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*:: type PatchChannel = {
  subscribe: () => () => {},
  receivePatches: (patches: Array<*>) => void
}*/
/*:: type Props = {
  value: ?any,
  schema: any,
  type: Object,
  patchChannel: PatchChannel,
  onChange: () => {},
  autoFocus: boolean
}*/
class SanityFormBuilder extends _react2.default.Component /*:: <Props>*/ {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.setInput = (input /*: ?FormBuilderInput*/) => {
      this._input = input;
    }, this.renderInput = ({ onFocus, onBlur, focusPath }) => {
      var _props = this.props;
      const value = _props.value,
            type = _props.type,
            onChange = _props.onChange;

      return _react2.default.createElement(_FormBuilderInput.FormBuilderInput, {
        type: type,
        onChange: onChange,
        level: 0,
        value: value,
        onFocus: onFocus,
        onBlur: onBlur,
        focusPath: focusPath,
        isRoot: true,
        ref: this.setInput
      });
    }, _temp;
  }

  componentDidMount() {
    const autoFocus = this.props.autoFocus;

    if (this._input && autoFocus) {
      this._input.focus();
    }
  }

  render() {
    var _props2 = this.props;
    const value = _props2.value,
          schema = _props2.schema,
          patchChannel = _props2.patchChannel;

    return _react2.default.createElement(
      _SanityFormBuilderContext2.default,
      {
        value: value,
        schema: schema,
        patchChannel: patchChannel
      },
      _react2.default.createElement(
        _SimpleFocusManager2.default,
        null,
        this.renderInput
      )
    );
  }
}

exports.default = SanityFormBuilder;
SanityFormBuilder.createPatchChannel = _SanityFormBuilderContext2.default.createPatchChannel;
